File: backend/src/main/kotlin/top/bearingwall/asya/service/ConferenceService.kt
package top.bearingwall.asya.service

import jakarta.persistence.criteria.JoinType
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.audit.Auditable
import top.bearingwall.asya.dto.ConferenceRequest
import top.bearingwall.asya.dto.ConferenceResponse
import top.bearingwall.asya.dto.UserInfoResponse
import top.bearingwall.asya.model.AuditActionType
import top.bearingwall.asya.model.Conference
import top.bearingwall.asya.model.ConferenceStatus
import top.bearingwall.asya.model.User
import top.bearingwall.asya.model.UserRole
import top.bearingwall.asya.repository.ConferenceRepository
import top.bearingwall.asya.repository.UserRepository
import java.util.UUID

@Service
class ConferenceService(
    private val conferenceRepository: ConferenceRepository,
    private val userRepository: UserRepository
) {
    @Transactional
    @Auditable(type = AuditActionType.CONFERENCE_CREATE, content = "创建会议")
    fun createConference(requester: User, req: ConferenceRequest): ConferenceResponse {
        require(requester.role == UserRole.SYS_ADMIN) { "Only SYS_ADMIN can create conference" }
        val conf = Conference(
            name = req.name,
            description = req.description,
            status = req.status ?: ConferenceStatus.PREPARING
        )
        val saved = conferenceRepository.save(conf)
        return saved.toResponse()
    }

    @Transactional
    @Auditable(type = AuditActionType.CONFERENCE_UPDATE, content = "更新会议")
    fun updateConference(requester: User, req: ConferenceRequest): ConferenceResponse {
        require(requester.role == UserRole.DH || requester.role == UserRole.DM || requester.role == UserRole.SYS_ADMIN) {
            "Only DH, DM, or SYS_ADMIN can update conference"
        }
        // TODO：检查请求者是否有权限修改该会议
        val user = userRepository.findById(requester.uuid!!).orElseThrow { IllegalStateException("User not found") }
        val conf = user.conference ?: throw IllegalStateException("Requester not associated with any conference")
        req.name.let { conf.name = it }
        req.description.let { conf.description = it }
        req.status?.let { conf.status = it }
        val saved = conferenceRepository.save(conf)
        return saved.toResponse()
    }

    @Transactional(readOnly = true)
    fun getMyConference(requester: User): ConferenceResponse {
        val user = userRepository.findById(requester.uuid!!).orElseThrow { IllegalStateException("User not found") }
        val conf = user.conference ?: throw IllegalStateException("Requester not associated with any conference")
        return conf.toResponse()
    }

    @Transactional(readOnly = true)
    fun getConferenceUsers(requester: User): List<UserInfoResponse> {
        val user = userRepository.findById(requester.uuid!!).orElseThrow { IllegalStateException("User not found") }
        val conf = user.conference ?: throw IllegalStateException("Requester not associated with any conference")
        return conf.users.map { u ->
            UserInfoResponse(
                uuid = u.uuid?.toString() ?: "",
                name = u.name,
                displayName = u.displayName,
                role = u.role,
                conferenceUuid = u.conference?.uuid?.toString(),
                conferenceName = u.conference?.name
            )
        }
    }

    @Transactional(readOnly = true)
    fun getConferenceDelegates(
        requester: User,
        pageable: Pageable,
        name: String?,
        displayName: String?
    ): Page<UserInfoResponse> {
        val user = userRepository.findById(requester.uuid!!).orElseThrow { IllegalStateException("User not found") }
        val conf = user.conference ?: throw IllegalStateException("Requester not associated with any conference")
        val conferenceUuid = conf.uuid!!

        val specification = Specification<User> { root, _, cb ->
            val predicates = mutableListOf<jakarta.persistence.criteria.Predicate>()

            val conference = root.join<User, Any>("conference", JoinType.LEFT)
            predicates += cb.equal(conference.get<UUID>("uuid"), conferenceUuid)
            predicates += cb.equal(root.get<UserRole>("role"), UserRole.DELEGATE)

            name?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(cb.lower(root.get("name")), "%${keyword.lowercase()}%")
            }

            displayName?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(cb.lower(root.get("displayName")), "%${keyword.lowercase()}%")
            }

            cb.and(*predicates.toTypedArray())
        }

        return userRepository.findAll(specification, pageable).map { u ->
            UserInfoResponse(
                uuid = u.uuid?.toString() ?: "",
                name = u.name,
                displayName = u.displayName,
                role = u.role,
                conferenceUuid = u.conference?.uuid?.toString(),
                conferenceName = u.conference?.name
            )
        }
    }

    @Transactional(readOnly = true)
    fun listAll(requester: User): List<ConferenceResponse> {
        require(requester.role == UserRole.SYS_ADMIN) { "Only SYS_ADMIN can list all conferences" }
        return conferenceRepository.findAll().map { it.toResponse() }
    }

    @Transactional(readOnly = true)
    fun listPage(requester: User, pageable: Pageable): Page<ConferenceResponse> {
        require(requester.role == UserRole.SYS_ADMIN) { "Only SYS_ADMIN can list conferences" }
        return conferenceRepository.findAll(pageable).map { it.toResponse() }
    }

    @Transactional
    @Auditable(type = AuditActionType.CONFERENCE_UPDATE, content = "管理员更新会议")
    fun updateConferenceByUuid(requester: User, conferenceUuid: UUID, req: ConferenceRequest): ConferenceResponse {
        require(requester.role == UserRole.SYS_ADMIN) { "Only SYS_ADMIN can update conference by uuid" }
        val conf = conferenceRepository.findById(conferenceUuid).orElseThrow {
            IllegalStateException("Conference not found: $conferenceUuid")
        }
        req.name.let { conf.name = it }
        req.description.let { conf.description = it }
        req.status?.let { conf.status = it }
        val saved = conferenceRepository.save(conf)
        return saved.toResponse()
    }

    @Transactional
    @Auditable(type = AuditActionType.CONFERENCE_ASSIGN_USER, content = "分配用户到会议")
    fun assignUserToConference(requester: User, conferenceUuid: UUID, userUuid: UUID): UserInfoResponse {
        require(requester.role == UserRole.SYS_ADMIN) { "Only SYS_ADMIN can assign users to conference" }
        val conference = conferenceRepository.findById(conferenceUuid).orElseThrow {
            IllegalStateException("Conference not found")
        }
        val user = userRepository.findById(userUuid).orElseThrow {
            IllegalStateException("User not found")
        }
        user.conference = conference
        val saved = userRepository.save(user)
        return UserInfoResponse(
            uuid = saved.uuid?.toString() ?: "",
            name = saved.name,
            displayName = saved.displayName,
            role = saved.role,
            conferenceUuid = saved.conference?.uuid?.toString(),
            conferenceName = saved.conference?.name
        )
    }

    @Transactional(readOnly = true)
    fun isInstructionSubmissionPaused(conferenceUuid: UUID): Boolean {
        val conference = conferenceRepository.findById(conferenceUuid)
            .orElseThrow { IllegalArgumentException("Conference not found") }
        return conference.instructionSubmissionPaused
    }

    @Transactional
    @Auditable(type = AuditActionType.INSTRUCTION_SUBMISSION_SWITCH, content = "设置会议指令提交暂停开关")
    fun setInstructionSubmissionPaused(conferenceUuid: UUID, paused: Boolean): ConferenceResponse {
        val conference = conferenceRepository.findById(conferenceUuid)
            .orElseThrow { IllegalArgumentException("Conference not found") }
        conference.instructionSubmissionPaused = paused
        val saved = conferenceRepository.save(conference)
        return saved.toResponse()
    }

    private fun Conference.toResponse(): ConferenceResponse = ConferenceResponse(
        uuid = this.uuid?.toString() ?: "",
        name = this.name,
        description = this.description,
        status = this.status,
        instructionSubmissionPaused = this.instructionSubmissionPaused
    )
}

File: backend-dotnet/src/AsyaMun.Api/Services/ConferenceService.cs
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public class ConferenceService
{
    private readonly AppDbContext _db;

    public ConferenceService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<ConferenceResponse> CreateConferenceAsync(User requester, ConferenceRequest request)
    {
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only SYS_ADMIN can create conference");
        }

        var conference = new Conference
        {
            Uuid = Guid.NewGuid(),
            Name = request.Name,
            Description = request.Description,
            Status = request.Status ?? ConferenceStatus.PREPARING
        };

        _db.Conferences.Add(conference);
        await _db.SaveChangesAsync();

        return ToResponse(conference);
    }

    public async Task<ConferenceResponse> UpdateConferenceAsync(User requester, ConferenceRequest request)
    {
        if (requester.Role is not (UserRole.DH or UserRole.DM or UserRole.SYS_ADMIN))
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only DH, DM, or SYS_ADMIN can update conference");
        }

        var user = await LoadRequesterAsync(requester);
        var conference = user.Conference
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Requester not associated with any conference");

        conference.Name = request.Name;
        conference.Description = request.Description;
        if (request.Status.HasValue)
        {
            conference.Status = request.Status.Value;
        }

        await _db.SaveChangesAsync();

        return ToResponse(conference);
    }

    public async Task<ConferenceResponse> GetMyConferenceAsync(User requester, CancellationToken ct = default)
    {
        var user = await LoadRequesterAsync(requester, ct);
        var conference = user.Conference
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Requester not associated with any conference");

        return ToResponse(conference);
    }

    public async Task<List<UserInfoResponse>> GetConferenceUsersAsync(User requester, CancellationToken ct = default)
    {
        var user = await LoadRequesterAsync(requester, ct);
        var conference = user.Conference
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Requester not associated with any conference");

        var members = await _db.Users
            .AsNoTracking()
            .Include(u => u.Conference)
            .Where(u => u.ConferenceId == conference.Uuid)
            .ToListAsync(ct);

        return members.Select(UserInfoResponse.From).ToList();
    }

    public async Task<Page<UserInfoResponse>> GetConferenceDelegatesAsync(
        User requester,
        PageInput input,
        string? name,
        string? displayName,
        CancellationToken ct = default)
    {
        var user = await LoadRequesterAsync(requester, ct);
        var conference = user.Conference
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Requester not associated with any conference");

        IQueryable<User> query = _db.Users
            .AsNoTracking()
            .Include(u => u.Conference)
            .Where(u => u.ConferenceId == conference.Uuid && u.Role == UserRole.DELEGATE);

        if (!string.IsNullOrWhiteSpace(name))
        {
            var keyword = name.Trim().ToLower();
            query = query.Where(u => u.Name.ToLower().Contains(keyword));
        }

        if (!string.IsNullOrWhiteSpace(displayName))
        {
            var keyword = displayName.Trim().ToLower();
            query = query.Where(u => u.DisplayName != null && u.DisplayName.ToLower().Contains(keyword));
        }

        var page = await query.ToPageAsync(input, ct);

        return Page<UserInfoResponse>.Of(
            page.Content.Select(UserInfoResponse.From).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task<List<ConferenceResponse>> ListAllAsync(User requester, CancellationToken ct = default)
    {
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only SYS_ADMIN can list all conferences");
        }

        var conferences = await _db.Conferences.AsNoTracking().ToListAsync(ct);

        return conferences.Select(ToResponse).ToList();
    }

    public async Task<Page<ConferenceResponse>> ListPageAsync(
        User requester,
        PageInput input,
        CancellationToken ct = default)
    {
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only SYS_ADMIN can list conferences");
        }

        var page = await _db.Conferences.AsNoTracking().ToPageAsync(input, ct);

        return Page<ConferenceResponse>.Of(
            page.Content.Select(ToResponse).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task<ConferenceResponse> UpdateConferenceByUuidAsync(
        User requester,
        Guid conferenceUuid,
        ConferenceRequest request,
        CancellationToken ct = default)
    {
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only SYS_ADMIN can update conference by uuid");
        }

        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceUuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Conference not found: {conferenceUuid}");

        conference.Name = request.Name;
        conference.Description = request.Description;
        if (request.Status.HasValue)
        {
            conference.Status = request.Status.Value;
        }

        await _db.SaveChangesAsync(ct);

        return ToResponse(conference);
    }

    public async Task<UserInfoResponse> AssignUserToConferenceAsync(
        User requester,
        Guid conferenceUuid,
        Guid userUuid,
        CancellationToken ct = default)
    {
        if (requester.Role != UserRole.SYS_ADMIN)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only SYS_ADMIN can assign users to conference");
        }

        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceUuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Conference not found");

        var user = await _db.Users.Include(u => u.Conference).FirstOrDefaultAsync(u => u.Uuid == userUuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "User not found");

        user.ConferenceId = conference.Uuid;
        user.Conference = conference;

        await _db.SaveChangesAsync(ct);

        return UserInfoResponse.From(user);
    }

    public async Task<bool> IsInstructionSubmissionPausedAsync(Guid conferenceUuid, CancellationToken ct = default)
    {
        var conference = await _db.Conferences.AsNoTracking().FirstOrDefaultAsync(c => c.Uuid == conferenceUuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Conference not found");

        return conference.InstructionSubmissionPaused;
    }

    public async Task<ConferenceResponse> SetInstructionSubmissionPausedAsync(
        Guid conferenceUuid,
        bool paused,
        CancellationToken ct = default)
    {
        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceUuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Conference not found");

        conference.InstructionSubmissionPaused = paused;

        await _db.SaveChangesAsync(ct);

        return ToResponse(conference);
    }

    private async Task<User> LoadRequesterAsync(User requester, CancellationToken ct = default)
    {
        return await _db.Users.Include(u => u.Conference)
            .FirstOrDefaultAsync(u => u.Uuid == requester.Uuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "User not found");
    }

    private static ConferenceResponse ToResponse(Conference conference)
    {
        return new ConferenceResponse(
            conference.Uuid.ToString("D"),
            conference.Name,
            conference.Description,
            conference.Status,
            conference.InstructionSubmissionPaused);
    }
}